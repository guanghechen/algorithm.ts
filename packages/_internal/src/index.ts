export * from './constant'
export * from './types'
