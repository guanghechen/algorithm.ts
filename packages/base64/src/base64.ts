export interface IBase64Props {
  /**
   * @default 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/'
   */
  CODES?: string
  /**
   * @default '=''
   */
  CODE_PADDING?: string
}

const sanitizeRegex = /([*^$\\])/g

export class Base64 {
  protected readonly CODE_LIST: string[]
  protected readonly CODE_REFLECT: Record<string, number>
  protected readonly CODE_PADDING: string
  protected readonly VALIDATE_REGEX: RegExp

  constructor(props: IBase64Props = {}) {
    const CODES =
      props.CODES?.length === 64
        ? props.CODES
        : 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/'
    const CODE_PADDING = props.CODE_PADDING?.length === 1 ? props.CODE_PADDING : '='

    this.CODE_LIST = CODES.split('')
    this.CODE_REFLECT = this.CODE_LIST.reduce((acc, v, idx) => ({ ...acc, [v]: idx }), {
      [CODE_PADDING]: 0,
    })
    this.VALIDATE_REGEX = new RegExp(
      `^[${CODES.replace(sanitizeRegex, '\\$1')}]+${CODE_PADDING.replace(
        sanitizeRegex,
        '\\$1',
      )}{0,2}$`,
    )
    this.CODE_PADDING = CODE_PADDING
  }

  public encode(data: Uint8Array): string {
    let result = ''
    const _end: number = Math.floor(data.length / 3) * 3
    for (let i = 0; i < _end; i += 3) {
      result += this._encodeUnit((data[i] << 16) | (data[i + 1] << 8) | data[i + 2])
    }

    const { CODE_PADDING } = this
    switch (data.length - _end) {
      case 1: {
        const tail: string = this._encodeUnit(data[_end] << 16)
        result += tail.slice(0, 2) + CODE_PADDING + CODE_PADDING
        break
      }
      case 2: {
        const tail: string = this._encodeUnit((data[_end] << 16) | (data[_end + 1] << 8))
        result += tail.slice(0, 3) + CODE_PADDING
      }
    }
    return result
  }

  public decode(text: string): Uint8Array | never {
    if (text.length <= 0 || !this.validate(text)) {
      throw new TypeError('Invalid base64 string.')
    }

    let countOfPadding = 0
    const { CODE_PADDING } = this
    for (let i = text.length - 1; i >= 0 && text[i] === CODE_PADDING; --i) countOfPadding += 1
    const _size: number = (text.length >> 2) * 3 - countOfPadding
    const result: Uint8Array = new Uint8Array(_size)

    const _end: number = ((text.length - countOfPadding) >> 2) << 2
    let i = 0
    let j = 0
    for (; i < _end; i += 4, j += 3) {
      const v: number = this._decodeUint(text, i)
      result[j] = v >> 16
      result[j + 1] = (v >> 8) & 0xff
      result[j + 2] = v & 0xff
    }

    if (countOfPadding > 0) {
      const v: number = this._decodeUint(text, i)
      switch (countOfPadding) {
        case 1: {
          result[j] = v >> 16
          result[j + 1] = (v >> 8) & 0xff
          break
        }
        case 2: {
          result[j] = v >> 16
          break
        }
      }
    }
    return result
  }

  public validate(text: string): boolean {
    // The length of a Base64 encoded string should be a multiple of 4.
    if (text.length & 3) return false
    return this.VALIDATE_REGEX.test(text)
  }

  protected _encodeUnit(v: number): string {
    const { CODE_LIST } = this
    return (
      CODE_LIST[(v >> 18) & 0x3f] +
      CODE_LIST[(v >> 12) & 0x3f] +
      CODE_LIST[(v >> 6) & 0x3f] +
      CODE_LIST[v & 0x3f]
    )
  }

  protected _decodeUint(text: string, startIdx: number): number {
    const { CODE_REFLECT } = this
    const v: number =
      (CODE_REFLECT[text.charAt(startIdx)] << 18) |
      (CODE_REFLECT[text.charAt(startIdx | 1)] << 12) |
      (CODE_REFLECT[text.charAt(startIdx | 2)] << 6) |
      CODE_REFLECT[text.charAt(startIdx | 3)]
    return v
  }
}
