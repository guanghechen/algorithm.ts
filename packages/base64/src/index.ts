const BASE64_CODES = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/'
const BASE64_CODE_PADDING = '='
const BASE64_CODE_LIST = BASE64_CODES.split('')
const BASE64_CODE_MAP = BASE64_CODE_LIST.reduce((acc, v, idx) => ({ ...acc, [v]: idx }), {
  [BASE64_CODE_PADDING]: 0,
})
const BASE64_VALIDATE_REGEX = new RegExp(`^[${BASE64_CODES}]+={0,2}$`)

export default { encode, decode, validate }

export function encode(data: Uint8Array): string {
  let result = ''
  const _end: number = Math.floor(data.length / 3) * 3
  for (let i = 0; i < _end; i += 3) {
    result += encodeUnit((data[i] << 16) | (data[i + 1] << 8) | data[i + 2])
  }

  switch (data.length - _end) {
    case 1: {
      const tail: string = encodeUnit(data[_end] << 16)
      result += tail.slice(0, 2) + BASE64_CODE_PADDING + BASE64_CODE_PADDING
      break
    }
    case 2: {
      const tail: string = encodeUnit((data[_end] << 16) | (data[_end + 1] << 8))
      result += tail.slice(0, 3) + BASE64_CODE_PADDING
    }
  }
  return result
}

export function decode(text: string): Uint8Array | never {
  if (text.length <= 0 || !validate(text)) {
    throw new TypeError('Invalid base64 string.')
  }

  let countOfPadding = 0
  for (let i = text.length - 1; i >= 0 && text[i] === BASE64_CODE_PADDING; --i) countOfPadding += 1
  const _size: number = (text.length >> 2) * 3 - countOfPadding
  const result: Uint8Array = new Uint8Array(_size)

  const _end: number = ((text.length - countOfPadding) >> 2) << 2
  let i = 0
  let j = 0
  for (; i < _end; i += 4, j += 3) {
    const v: number = decodeUint(text, i)
    result[j] = v >> 16
    result[j + 1] = (v >> 8) & 0xff
    result[j + 2] = v & 0xff
  }

  if (countOfPadding > 0) {
    const v: number = decodeUint(text, i)
    switch (countOfPadding) {
      case 1: {
        result[j] = v >> 16
        result[j + 1] = (v >> 8) & 0xff
        break
      }
      case 2: {
        result[j] = v >> 16
        break
      }
    }
  }
  return result
}

export function validate(text: string): boolean {
  // The length of a Base64 encoded string should be a multiple of 4.
  if (text.length & 3) return false
  return BASE64_VALIDATE_REGEX.test(text)
}

function encodeUnit(v: number): string {
  return (
    BASE64_CODE_LIST[(v >> 18) & 0x3f] +
    BASE64_CODE_LIST[(v >> 12) & 0x3f] +
    BASE64_CODE_LIST[(v >> 6) & 0x3f] +
    BASE64_CODE_LIST[v & 0x3f]
  )
}

function decodeUint(text: string, startIdx: number): number {
  const v: number =
    (BASE64_CODE_MAP[text.charAt(startIdx)] << 18) |
    (BASE64_CODE_MAP[text.charAt(startIdx | 1)] << 12) |
    (BASE64_CODE_MAP[text.charAt(startIdx | 2)] << 6) |
    BASE64_CODE_MAP[text.charAt(startIdx | 3)]
  return v
}
