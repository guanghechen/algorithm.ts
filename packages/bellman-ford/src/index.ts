import { createCircularQueue } from '@algorithm.ts/circular-queue'

export interface IEdge {
  /**
   * The other end of the edge.
   */
  to: number
  /**
   * The cost of walking along this side.
   */
  cost: number
}

export interface IGraph {
  /**
   * The number of nodes in the graph. (0-index)
   */
  N: number
  /**
   * The source node. (0-index)
   */
  source: number
  /**
   * Graph edges.
   */
  edges: ReadonlyArray<IEdge>
  /**
   * Adjacency list. G[i] represent the index list of the  edges start from node i.
   */
  G: ReadonlyArray<ReadonlyArray<number>>
  /**
   * An array recording the shortest distance to the source point.
   */
  dist: number[]
}

export interface IOptions {
  /**
   * A big number, representing the unreachable cost.
   */
  INF?: number
  /**
   * Used to check if an element is already in the queue.
   */
  inq?: boolean[]
  /**
   * Record the number of times an element is enqueued,
   * used to check whether there is a negative cycle.
   */
  inqTimes?: number[]
}

const ZERO = 0
const DEFAULT_INF = Math.floor(Number.MAX_SAFE_INTEGER / 2)
const Q = createCircularQueue<number>()

/**
 * The bellman-ford algorithm, optimized with queue.
 *
 * @param graph
 * @param options
 */
export function bellmanFord(graph: IGraph, options: IOptions = {}): boolean {
  const { N, source, edges, G, dist } = graph
  const { INF = DEFAULT_INF, inq = [], inqTimes = [] } = options

  if (dist.length < N) dist.length = N
  if (inq.length < N) inq.length = N
  if (inqTimes.length < N) inqTimes.length = N

  dist.fill(INF, 0, N)
  inq.fill(false, 0, N)
  inqTimes.fill(0, 0, N)

  Q.init(N + 1)
  Q.enqueue(source)
  dist[source] = ZERO

  while (Q.size() > 0) {
    const o: number = Q.dequeue()!
    inq[o] = false

    for (const idx of G[o]) {
      const edge: IEdge = edges[idx]
      if (dist[edge.to] > dist[o] + edge.cost) {
        dist[edge.to] = dist[o] + edge.cost
        if (!inq[edge.to]) {
          Q.enqueue(edge.to)
          inq[edge.to] = true

          // eslint-disable-next-line no-plusplus
          if (++inqTimes[edge.to] > N) return false
        }
      }
    }
  }
  return true
}

export default bellmanFord
