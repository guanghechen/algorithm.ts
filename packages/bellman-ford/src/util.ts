/**
 * List all points on the shortest path from source to target in order.
 *
 * @param bestFrom
 * @param source
 * @param target
 * @returns
 */
export const getShortestPath = (
  bestFrom: ReadonlyArray<number>,
  source: number,
  target: number,
): number[] => {
  const path: number[] = [target]
  for (let x = target, parent: number; x !== source; x = parent) {
    parent = bestFrom[x]
    path.push(parent)
  }
  return path.reverse()
}
