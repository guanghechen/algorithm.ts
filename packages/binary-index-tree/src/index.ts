export * from './tree1'
export * from './tree2'
export * from './types'
export * from './util'
