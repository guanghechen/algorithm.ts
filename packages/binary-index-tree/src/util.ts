/**
 * Calculate the minimum bit value of x
 *
 * @param x
 */
export function lowbit(x: number): number {
  return x & -x
}
