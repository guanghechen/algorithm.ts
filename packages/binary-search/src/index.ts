export * from './binary-search'
export * from './lower-bound'
export * from './upper-bound'
export * from './types'
