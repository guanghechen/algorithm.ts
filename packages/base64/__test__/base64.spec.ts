import { TextEncoder } from 'util'
import { decode, encode, validate } from '../src'

describe('basic', function () {
  test('%4 == 0', () => testWithBytes(1000))
  test('%4 == 1', () => testWithBytes(1001))
  test('%4 == 2', () => testWithBytes(1002))
  test('%4 == 3', () => testWithBytes(1003))

  function testWithBytes(size: number): void {
    const data = new Uint8Array(size)
    for (let i = 0; i < size; ++i) data[i] = Math.round(Math.random() * 256) * 0xff

    const encodedData = encode(data)
    expect(typeof encodedData).toBe('string')
    expect(validate(encodedData)).toBe(true)

    const decodedData = decode(encodedData)
    expect(decodedData).toBeInstanceOf(Uint8Array)
    expect(data.length).toEqual(decodedData.length)

    expect(data).toEqual(decodedData)
  }
})

describe('unexpected', function () {
  test('bad base64 string', function () {
    expect(validate('a')).toBe(false)
    expect(validate('aa')).toBe(false)
    expect(validate('aaa')).toBe(false)
    expect(validate('a===')).toBe(false)

    expect(() => decode('a')).toThrow(/Invalid base64 string/)
  })
})

test('snapshot', function () {
  const textEncoder = new TextEncoder()
  const getBytes = (text: string): Uint8Array => textEncoder.encode(text)

  expect(encode(getBytes('Hello, world!'))).toMatchSnapshot()
  expect(encode(getBytes('Hello, world!1'))).toMatchSnapshot()
  expect(encode(getBytes('Hello, world!11'))).toMatchSnapshot()
  expect(encode(getBytes('Hello, world!111'))).toMatchSnapshot()
})
