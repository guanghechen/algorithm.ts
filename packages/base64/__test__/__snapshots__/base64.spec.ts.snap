// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot 1`] = `"SGVsbG8sIHdvcmxkIQ=="`;

exports[`snapshot 2`] = `"SGVsbG8sIHdvcmxkITE="`;

exports[`snapshot 3`] = `"SGVsbG8sIHdvcmxkITEx"`;

exports[`snapshot 4`] = `"SGVsbG8sIHdvcmxkITExMQ=="`;
